### ISSUE TITLE
[feat] Tracking issue MQTTv5

### ISSUE THREAD

**Issue by robertsLando** [moscajs/aedes#821]
The features that doesn't require changes on persistences are: 

- [ ]  [Reason code](https://docs.oasis-open.org/mqtt/mqtt/v5.0/os/mqtt-v5.0-os.html#_Toc3901031) on all ACKs
- [ ]  Reason string on all ACKs
- [ ]  Message expiry: the lifetime of the Will Message in seconds
- [ ]  Server disconnect
- [ ]  Payload format and content type (this should be already handled by mqtt-packet we just have to forward it)
- [ ]  [Request / Response](https://docs.oasis-open.org/mqtt/mqtt/v5.0/os/mqtt-v5.0-os.html#_Toc3901252)
- [ ]  Subscription id (persistence `addSubscription` stores the entire sub packet so we don't have to make edits on that side)
- [ ]  Flow control
- [ ]  User properties (this should be already handled by mqtt-packet we just have to forward it)
- [ ]  Maximum Packet Size
- [ ]  Optional Server feature availability
- [ ]  Enhanced authentication
- [ ]  [Subscription options](https://docs.oasis-open.org/mqtt/mqtt/v5.0/os/mqtt-v5.0-os.html#_Toc3901169): I think this should already be implemented, expecially the noLocal flag but we should check this
- [ ]  Will delay: checking the `putWill` we store the entire packet in persistence so no change is needed on that part
- [ ]  Server Keep Alive
- [ ]  Assigned ClientID
- [ ]  Server reference

The ones that require changes on persistences:

- [ ] Session expiry: we actually don't have storage support for this
- [ ] Shared subscriptions: I drafted an implementation in #449 but i never ended it
- [ ] Topic alias

**Comment by robertsLando** [comment]
✅ **#1088 merged into `mqttv5`** on 2026-06-30 (merge commit `e7d0ce9`).

- Companion releases are out, so `package.json` now uses published ranges (`aedes-packet@^4.0.1`, `aedes-persistence@^10.3.1`); the `overrides` shim and the git-dep release block are gone.
- Because #1088 targeted `mqttv5` (not the default branch), the per-feature issues it referenced were **not** auto-closed — the checklist at the top continues to track them until `mqttv5` lands for release under #194.

The development-status section above has been updated to reflect the merge.

**Comment by BenjaminDobler** [comment]
## Proposed sequence for the remaining **no-persistence** items

Per @robertsLando's suggestion (start with the ones that don't touch persistence). Grouped into reviewable PRs against `mqttv5`, ordered by conformance value × independence. All of these are broker-core only and cluster-safe.

### PR 1 — Reason codes & reason strings on ACKs (#822, #823, #1092)
- **#823** add `reasonString` to ACK packets (PUBACK / PUBREC / PUBREL / PUBCOMP / SUBACK / UNSUBACK / server-DISCONNECT).
- **#1092** honor `requestProblemInformation = 0` → suppress `reasonString` **and** `userProperties` on every packet except PUBLISH / CONNACK / DISCONNECT ([MQTT-3.1.2-29]).
- **#822** granular reason codes: PUBREL/PUBCOMP `0x00`; per-topic SUBACK failure codes instead of the coarse `0x80`; PUBACK/PUBREC codes beyond `0x87`.
- _Why together:_ all three edit the same ACK builders, and #1092 is the gate for #823 — splitting them would churn the same files three times. Can peel #822 into its own PR if the diff gets large.

### PR 2 — Request / Response Information (#1091)
CONNECT `0x19` → CONNACK `responseInformation` `0x1A`. Small and standalone; optional (server MAY, §3.2.2.3.15) but cheap. Needs a small operator hook (broker option/callback) to supply the response-topic base.

### PR 3 — Server Reference (#838)
`serverReference` (`0x1C`) on CONNACK (`0x9C` use-another-server) and DISCONNECT (`0x9C`/`0x9D` server-moved). DISCONNECT already carries it via `client.disconnect({ reasonCode, properties: { serverReference } })`; the new part is a tiny reject-with-reference path at connect.

### Deferred (happy to propose separately when you want them)
- **#833** full enhanced-auth AUTH packet flow (re-auth `0x18`/`0x19`) — larger, design-heavy.
- **#829** outbound Receive-Maximum enforcement — invasive to the shared delivery hot path.
- **#840** outbound (broker-assigned) Topic Alias — optional optimization (inbound is done).
- **#828** multiple matching Subscription Identifiers — touches the dedupe/delivery core (documented limitation today).
- Persistence/cluster: **#839** cross-restart session expiry, **#841** cluster-aware shared subs, **#835** durable will delay.

Does this order work? I'd start with **PR 1**. I'll align each PR with your mqttv5 e2e compatibility workflow so the tests cover features as they land — and shout if you'd rather I sequence around what your e2e suite checks first.

**Comment by robertsLando** [comment]
Yes go

**Comment by BenjaminDobler** [comment]
## Next sequence — remaining MQTT 5.0 work

The no-persistence-first batch is all merged 🎉 (#1100, #1102, #1115). Proposing the next steps, grouped and ordered by value × independence × risk. Base is `mqttv5`; each item is its own reviewable PR.

### Phase A — close the remaining Paho v5 compat gaps (broker-core, cluster-safe)

Each of these currently sits in `EXPECTED_GAPS` in `tools/mqtt-compat/run_compat.py`; landing it flips a `⚠️ expected gap` into a pass and lifts the reported MQTT 5.0 compat %:

1. **#828 — multiple matching Subscription Identifiers** [MQTT-3.3.4-4] → `test_subscribe_identifiers`. Today a delivery matching several of a client's overlapping subscriptions echoes only one SI (the documented limitation); this delivers all matching ones. Touches the dedupe/delivery path, no persistence. __Suggest starting here__ — clearest conformance win, smallest blast radius.
2. **#840 — outbound (broker-assigned) Topic Alias** (server MAY) → `test_server_topic_alias`. Inbound aliases already work; this assigns aliases on outbound PUBLISH per connection. Per-connection state, moderate hot-path.
3. **#829 — outbound Receive Maximum enforcement** (§4.9) → `test_flow_control1`. Cap in-flight QoS 1/2 toward each client. The most invasive of the three (shared delivery hot path) — deferred originally for exactly that reason — so I'd do it last and benchmark it.

### Phase B — Enhanced Authentication (#833), independent feature

Full `AUTH`-packet flow (§4.12): CONNECT `Authentication Method` → `AUTH` (0x18 continue) exchange → CONNACK, plus re-auth (0x19). Today an auth-method CONNECT is rejected with `0x8C`. Needs a broker auth-exchange hook (SASL / token style). No persistence, cluster-safe, and not tied to a Paho compat test, so I've split it out. Higher design surface — happy to post a short hook-API proposal for sign-off before writing code.

### Phase C — persistence / cluster (multi-repo, later track)

These ripple into `aedes-persistence` + the mongodb/redis/level backends and/or need cluster coordination, so best treated as a separate track once A/B settle:

- **#839** cross-restart Session Expiry — persisted deadline + restart reaper
- **#835** durable Will Delay — persisted delay deadline + reaper
- **#841** cluster-aware Shared Subscriptions → `test_shared_subscriptions` — global group membership + deterministic recipient selection via `mqemitter-redis` / `-mongodb`

---

Does this order work? I'd kick off with **#828** unless you'd rather I clear all three Phase-A compat gaps first and leave #833 for after. For #833 I'll send a short hook-API sketch before implementing. 👍

**Comment by robertsLando** [comment]
I would work on: #823 #829 #833 #840 first and then start with the once that require persistence changes

**Comment by robertsLando** [comment]
let's clear **all** the no-persistence work before opening the persistence/cluster track. Two reasons it's the right cut: every such PR stays single-repo and cluster-safe (no `aedes-persistence` + mongodb/redis/level coordination), and each one flips a Paho `EXPECTED_GAPS` entry into a pass, so the reported MQTT 5.0 compat % keeps climbing as visible progress.

Worth noting the two plans barely disagree — @BenjaminDobler's Phase A + Phase B are already all broker-core/cluster-safe, and only Phase C touches persistence. The one thing to correct: the batch isn't quite "all merged" yet.

**#823 is still partial.** Reason String landed on **PUBACK/PUBREC only** (#1100); SUBACK/UNSUBACK/PUBREL/PUBCOMP still omit it. That's pure broker-core, no persistence — so it belongs in this batch

Reconciling both lists against the table, the complete no-persistence set is:

- **#823** — Reason String on the remaining ACK types (SUBACK/UNSUBACK/PUBREL/PUBCOMP)
- **#828** — multiple matching Subscription Identifiers [MQTT-3.3.4-4]. Storage already landed; the remaining multi-match fix is dedupe/delivery-path only, so it's core-only despite the "persist" tag on the row.
- **#829** — outbound Receive Maximum enforcement (§4.9)
- **#833** — full AUTH-packet flow (§4.12)
- **#840** — outbound (broker-assigned) Topic Alias

On ordering inside the batch I'd keep @BenjaminDobler's risk-based sequencing:

- **#829 last, with a benchmark** — it's the only one on the shared delivery hot path (deferred originally for exactly that reason).
- **#833 in parallel** — independent, higher design surface, and not gating a Paho compat test, so it shouldn't block the quick conformance wins. Happy to review the hook-API sketch first.
- **#828 / #840 / #823** are the clean, low-blast-radius compat-gap wins — good to start here.

Then **#839 / #835 / #841** as one coherent multi-repo persistence/cluster epic once the above settles. 👍

**Comment by robertsLando** [comment]
FYI aedes@2.0.0-beta.1 is out now 🎉
